apis/apiextensions/v1alpha1/composition_types.go

```diff
@@ -0,0 +1,267 @@
+	errConfigMissing       = func(s string) string { return fmt.Sprintf("given type %s requires configuration", s) }
+	errTransformWithType   = func(s string) string { return fmt.Sprintf("%s transform could not resolve", s) }
+)
+
+// CompositionSpec specifies the desired state of the definition.
+type CompositionSpec struct {
+	// From refers to the type that this composition is compatible. The values
+	// for the underlying resources will be fetched from the instances of the
+	// From.
+	// +immutable
+	From TypeReference `json:"from"`
+
+	// To is the list of target resources that make up the composition.
+	To []ComposedTemplate `json:"to"`
+}
+
+// TypeReference is used to refer to a type for declaring compatibility.
+type TypeReference struct {
+	// APIVersion of the type.
+	APIVersion string `json:"apiVersion"`
+
+	// Kind of the type.
+	Kind string `json:"kind"`
+}
+
+// ComposedTemplate is used to provide information about how the composed resource
+// should be processed.
+type ComposedTemplate struct {
+	// Base is the target resource that the patches will be applied on.
+	Base runtime.RawExtension `json:"base"`
+
+	// Patches will be applied as overlay to the base resource.
+	// +optional
+	Patches []Patch `json:"patches,omitempty"`
+
+	// ConnectionDetails lists the propagation secret keys from this target
+	// resource to the composition instance connection secret.
+	// +optional
+	ConnectionDetails []ConnectionDetail `json:"connectionDetails,omitempty"`
+}
+
+// Patch is used to patch the field on the base resource at ToFieldPath
+// after piping the value that is at FromFieldPath of the target resource through
+// transformers.
+type Patch struct {
+
+	// FromFieldPath is the path of the field on the upstream resource whose value
+	// to be used as input.
+	FromFieldPath string `json:"fromFieldPath"`
+
+	// ToFieldPath is the path of the field on the base resource whose value will
+	// be changed with the result of transforms. Leave empty if you'd like to
+	// propagate to the same path on the target resource.
+	// +optional
+	ToFieldPath string `json:"toFieldPath,omitempty"`
+
+	// Transforms are the list of functions that are used as a FIFO pipe for the
+	// input to be transformed.
+	// +optional
+	Transforms []Transform `json:"transforms,omitempty"`
+}
+
+// Apply runs transformers and patches the target resource.
+func (c *Patch) Apply(from, to runtime.Object) error {
+	fromMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(from)
+	if err != nil {
+		return err
+	}
+
+	in, err := fieldpath.Pave(fromMap).GetValue(c.FromFieldPath)
+	if err != nil {
+		return err
+	}
+	out := in
+	for i, f := range c.Transforms {
+		if out, err = f.Transform(out); err != nil {
+			return errors.Wrap(err, errTransformAtIndex(i))
+		}
+	}
+
+	if u, ok := to.(interface{ UnstructuredContent() map[string]interface{} }); ok {
+		return fieldpath.Pave(u.UnstructuredContent()).SetValue(c.ToFieldPath, out)
+	}
```

Comment:
@muvaf I think this is another instance where the `UnstructuredContent()` call is handled by the below `ToUnstructured` call.

Reply:
If it is unstructured we won't need to call `FromUnstructured` since the change is in-place. However, if it's not, we have to call `FromUnstructured` and that call does the conversion no matter what. So, it's more about whether `SetValue` operates on the actual object or the intermediary unstructured that will have to be converted back to the original struct.